### PR TITLE
Actions and Sources - Siteleaf

### LINKED ISSUE #3783: Siteleaf
### App
Siteleaf

### Triggers
* New Collection
* New Document
* New Page
* New Site

### Actions
* Create Document

### PR DESCRIPTION
**Pull request by vellames**
Closes #3783 

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4668"></a>

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/siteleaf/sources/new-collection/new-collection.mjs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    siteId: {
+      propDefinition: [
+        app,
+        "siteId",
+      ],
+    },
+  },
+  methods: {
+    ...base.methods,
+    async fetchEvents() {
```
Comment (GTFalcao) [inline components/siteleaf/sources/new-collection/new-collection.mjs:22]: The `fetchEvents()` method is nearly identical across the sources. I think we can reuse most of this code into a common method in the `base` file for the sources.

An individual `getData` or similar method to fetch `listCollections`, `listPages`, etc. and `getSummary` method would take care of the parts that differ between the sources.

**Review by GTFalcao — DISMISSED** [review]
Hi @vellames , thanks for your contribution!

I commented on a few things that could be improved for reusability and maintainability, following Pipedream's DRY standards.

File: components/siteleaf/siteleaf.app.mjs [hunk 2]
```diff
@@ -1,11 +1,106 @@
...
+        headers: this._getHeaders(),
+        auth: {
+          username: this._getApiKey(),
+          password: this._getApiSecret(),
+        },
+      };
+      return res;
+    },
+    async listSites(page, ctx = this) {
+      return axios(ctx, this._getAxiosParams({
```
[the remark this answers is not in the document]
Comment (vellames) [reply components/siteleaf/siteleaf.app.mjs:63]: I don't get it exactly. Are you suggesting wrapping `axios` in a function and then calling that function? Like a function that only calls another function?. Could you please provide an example?

[the remark this answers is not in the document]
Comment (GTFalcao) [reply components/siteleaf/siteleaf.app.mjs]: I mean the axios lib should only be used directly in a single place. I see you've done this in a new `_makeHttpRequest` which is exactly what I was thinking, so, looks good!

File: components/siteleaf/actions/create-document/create-document.mjs [hunk 3]
```diff
@@ -0,0 +1,100 @@
+import app from "../../siteleaf.app.mjs";
+import constants from "../common/constants.mjs";
+
+export default {
+  name: "Create Document",
+  description: "Create a new document. [See the docs here](https://learn.siteleaf.com/api/documents/#create-a-document)",
+  key: "siteleaf-create-document",
+  version: "0.1.1",
```
Comment (luancazarine) [inline components/siteleaf/actions/create-document/create-document.mjs:8]: isn't this the first version?
```suggestion
  version: "0.0.1",
```

File: components/siteleaf/sources/new-collection/new-collection.mjs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import base from "../common/base.mjs";
+import app from "../../siteleaf.app.mjs";
+
+export default {
+  ...base,
+  key: "siteleaf-new-collection",
+  type: "source",
+  name: "New Collection",
+  description: "Emit new event when a new collection is created",
```
Comment (luancazarine) [inline components/siteleaf/sources/new-collection/new-collection.mjs:9]: Could you put the docs link?

File: components/siteleaf/sources/new-document/new-document.mjs [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import app from "../../siteleaf.app.mjs";
+import base from "../common/base.mjs";
+
+export default {
+  ...base,
+  key: "siteleaf-new-document",
+  type: "source",
+  name: "New Document",
+  description: "Emit new event when a new document is created",
```
Comment (luancazarine) [inline components/siteleaf/sources/new-document/new-document.mjs:9]: Could you put the docs link?

File: components/siteleaf/sources/new-page/new-page.mjs [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import base from "../common/base.mjs";
+import app from "../../siteleaf.app.mjs";
+
+export default {
+  ...base,
+  key: "siteleaf-new-page",
+  type: "source",
+  name: "New Page",
+  description: "Emit new event when a new page is created",
```
Comment (luancazarine) [inline components/siteleaf/sources/new-page/new-page.mjs:9]: Could you put the docs link?

File: components/siteleaf/sources/new-site/new-site.mjs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import base from "../common/base.mjs";
+
+export default {
+  ...base,
+  key: "siteleaf-new-site",
+  type: "source",
+  name: "New Site",
+  description: "Emit new event when a new site is created",
```
Comment (luancazarine) [inline components/siteleaf/sources/new-site/new-site.mjs:8]: Could you put the docs link?

**Review by luancazarine — CHANGES_REQUESTED** [review]
Hi @vellames I've just added a few minor request changes.

**Review by luancazarine — APPROVED** [review]

**Pull request merged**